### PR TITLE
support execArgv when forking with child_process

### PR DESCRIPTION
**Pull request by paul-marechal**
Implement a workaround to convert opts.execArgv into
opts.env.NODE_OPTIONS when calling child_process.fork.

Note that this will most likely require you to package apps using
--no-bytecode because some node options will prevent you from running
the packaged bytecode with new options.

Fixes https://github.com/vercel/pkg/issues/334
Fixes https://github.com/vercel/pkg/issues/909

### CODE REVIEW

**Comment by marechal-p** [comment]
There might be a smarter way to go about this by modifying `pkg-fetch` Node patches, but it's a bit much for me to ramp up on... In the meantime, this workaround unblocked me enough to be able to package an application that would potentially rely on execArgv when forking.

**Review by erossignon — CHANGES_REQUESTED** [review]
a unit test would be needed to demonstrate the effect and improved behavior, and prevent future regression

**Comment by marechal-p** [comment]
@erossignon you are right, done!

File: prelude/bootstrap.js [hunk 1]
```diff
@@ -1630,6 +1631,13 @@ function payloadFileSync(pointer) {
     const opts = args[pos];
     if (!opts.env) opts.env = require('util')._extend({}, process.env);
     if (opts.env.PKG_EXECPATH === 'PKG_INVOKE_NODEJS') return;
+    if (Array.isArray(opts.execArgv)) {
+      // fork: The patched Node.js binary prevents us from passing execArgv on
+      // the cli. Instead we'll pass those options through the environment.
+      // https://nodejs.org/docs/latest/api/cli.html#cli_node_options_options
+      opts.env.NODE_OPTIONS = opts.execArgv.join(' ');
+      delete opts.execArgv;
+    }
```
Comment (marechal-p) [inline prelude/bootstrap.js:1640]: First, I know this is a workaround so I don't know how welcome that kind of change is.

Second, would it make sense to move that logic right above the `PKG_INVOKE_NODEJS` check?

Comment (marechal-p) [inline prelude/bootstrap.js:1640]: First, I know this is a workaround so I don't know how welcome that kind of change is.

Second, would it make sense to move that logic right above the `PKG_INVOKE_NODEJS` check?

**Comment by marechal-p** [comment]
I noticed the tests were failing because I wrongfully guessed Node's default memory limits. I now use `--disallow-code-generation-from-strings` to check if Node options are correctly passed and used by the child runtime.

**Comment by marechal-p** [comment]
@robertsLando thanks for merging master into this PR. I rebased on master instead and force pushed.

I don't understand why the tests are behaving differently on the CI compared to my machine?

**Comment by marechal-p** [comment]
I removed the bit that failed on CI, it was checking the output but it's not really necessary as long as we already detect that something wrong happened by checking for exit code 4.

Hopefully it works on the CI now...

**Comment by marechal-p** [comment]
CI was green, I just rebased on master.

**Comment by robertsLando** [comment]
Let's see tests result

**Comment by hacke2** [comment]
When can merged, thank you

**Comment by paul-marechal** [comment]
I rebased on master and fixed merge conflicts.

**Comment by paul-marechal** [comment]
@erossignon @robertsLando bump

**Comment by robertsLando** [comment]
Let's see hoe test goes first

**Comment by SBell6hf** [comment]
This PR also fixes Issue https://github.com/vercel/pkg/issues/1438. I would be grateful if it's merged.

File: lib/packer.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ function hasAnyStore(record: FileRecord) {
 interface PackerOptions {
   records: FileRecords;
   entrypoint: string;
-  bytecode: string;
+  bytecode: boolean;
```
Comment (robertsLando) [inline lib/packer.ts:50]: Why this change?

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by git9am** [comment]
This PR looks helpful. Any chance to merge it?

**Comment by paul-marechal** [comment]
> Why this change?

@robertsLando it was a left-over of when I was trying to understand the code to get this PR to work.

File: lib/packer.ts (hunk 2 above)
Comment (paul-marechal) [reply lib/packer.ts]: It was a left-over of when I was trying to understand the code to get this PR to work.

**Comment by git9am** [comment]
Similar issues exist while calling `child_process.spawn` to start another node process with arguments.

Minimal code snippets to reproduce

```js
const { spawn } = require('child_process');

const node = spawn('node', ['--version']);
node.stdout.on('data', data => console.log(`[node] stdout: ${data}`));
node.stderr.on('data', data => console.log(`[node] stderr: ${data}`));
```

**Comment by paul-marechal** [comment]
@git9am I guess this is because `bootstrap.js` patches `spawn` to interpret `node` as the current packaged binary.

Though I have no clue how to handle this, because there's no easy way to know what's an execArgv and what's not from the argument list passed :/
